cardano-cli/src/Cardano/CLI/EraBased/Script/Spend/Types.hs

```diff
@@ -0,0 +1,82 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+
+module Cardano.CLI.EraBased.Script.Spend.Types
+  ( CliSpendScriptRequirements (..)
+  , PlutusRefScriptCliArgs (..)
+  , SimpleOrPlutusScriptCliArgs (..)
+  , ScriptDatumOrFileSpending (..)
+  , SimpleRefScriptCliArgs (..)
+  , SpendScriptWitness (..)
+  , createSimpleOrPlutusScriptFromCliArgs
+  , createPlutusReferenceScriptFromCliArgs
+  , createSimpleReferenceScriptFromCliArgs
+  )
+where
+
+import           Cardano.Api
+
+import           Cardano.CLI.Types.Common (ScriptDataOrFile)
+
+newtype SpendScriptWitness era
+  = SpendScriptWitness {sswScriptWitness :: ScriptWitness WitCtxTxIn era}
+  deriving Show
+
+data CliSpendScriptRequirements
+  = OnDiskSimpleOrPlutusScript SimpleOrPlutusScriptCliArgs
+  | OnDiskSimpleRefScript SimpleRefScriptCliArgs
+  | OnDiskPlutusRefScript PlutusRefScriptCliArgs
+  deriving Show
+
+data SimpleOrPlutusScriptCliArgs
+  = OnDiskPlutusScriptCliArgs
```

Comment:
There are similar types in `Cardano.CLI.EraBased.Script.Mint.Types`. I would suggest picking one naming convention and use it everywhere:
1. Insert purpose into the type name like `CliSpendScriptRequirements`
2. Use clashing names like: `SimpleOrPlutusScriptCliArgs` and expect users to import the module qualified

Now we have both which results in oddities like: `PlutusSpend.ScriptDatumOrFileSpending` - you have `spend` two times in the full type name, which just adds noise.

Reply:
I'd rather not go through trouble of doing this because these are going to be short lived types as I need to find a better way of representing the different plutus script types. Left defer this to after I have removed `ScriptWitnessFiles` because I'm an unsure as to which way I want to go at the moment.

Reply:
I'm ok with this. But let's not lose track of this. Othewise it'll just add up to the rest of the technical debt.